### PR TITLE
xdp loader features

### PR DESCRIPTION
**Pull request by LorenzoBianconi**
Add features command to xdp-loader in order to dump XDP features supported by the NIC

### CODE REVIEW

File: xdp-loader/xdp-loader.c [hunk 1]
```diff
@@ -370,6 +371,62 @@ int do_clean(const void *cfg, __unused const char *pin_root_path)
...
+	printf("NETDEV_XDP_ACT_RX_SG:\t\t%s\n",
+	       CHECK_XDP_FEATURE(NETDEV_XDP_ACT_RX_SG));
+	printf("NETDEV_XDP_ACT_NDO_XMIT_SG:\t%s\n",
+	       CHECK_XDP_FEATURE(NETDEV_XDP_ACT_NDO_XMIT_SG));
+
+	return 0;
+#else
+	__unused const void *i = iface;
+
+	return -EOPNOTSUPP;
```
Comment (tohojo) [inline xdp-loader/xdp-loader.c:419]: We should print a friendly error message here, not just return an error code. Same thing when the libbpf feature is there, but the kernel doesn't support querying features...

File: xdp-loader/xdp-loader.c [hunk 2]
```diff
@@ -370,6 +371,62 @@ int do_clean(const void *cfg, __unused const char *pin_root_path)
...
 }
 
+static const struct featuresopt {
+	struct iface iface;
+} defaults_features = {};
+
+static struct prog_option features_options[] = {
+	DEFINE_OPTION("dev", OPT_IFNAME, struct featuresopt, iface,
+		      .positional = true,
+		      .metavar = "[ifname]",
```
Comment (tohojo) [inline xdp-loader/xdp-loader.c:381]: This should be `<ifname>`, since the parameter is not optional...

File: xdp-loader/xdp-loader.c [hunk 3]
```diff
@@ -370,6 +371,62 @@ int do_clean(const void *cfg, __unused const char *pin_root_path)
...
+		      .required = true,
+		      .help = "Show XDP features for device [ifname]"),
```
Comment (tohojo) [inline xdp-loader/xdp-loader.c:383]: ...also here

File: xdp-loader/xdp-loader.8 [hunk 4]
```diff
@@ -160,6 +161,18 @@ Enable debug logging. Specify twice for even more verbosity.
 .PP
 Display a summary of the available options
 
+.SH "The FEATURES command"
+.PP
+The \fIfeatures\fP command displays the XDP features supported by the NIC.
```
Comment (tohojo) [inline xdp-loader/xdp-loader.8:166]: We should also document what the possible features are, and what they mean.

File: xdp-loader/xdp-loader.c [hunk 5]
```diff
@@ -370,6 +371,62 @@ int do_clean(const void *cfg, __unused const char *pin_root_path)
...
+	printf("NETDEV_XDP_ACT_NDO_XMIT:\t%s\n",
+	       CHECK_XDP_FEATURE(NETDEV_XDP_ACT_NDO_XMIT));
+	printf("NETDEV_XDP_ACT_XSK_ZEROCOPY:\t%s\n",
+	       CHECK_XDP_FEATURE(NETDEV_XDP_ACT_XSK_ZEROCOPY));
+	printf("NETDEV_XDP_ACT_HW_OFFLOAD:\t%s\n",
+	       CHECK_XDP_FEATURE(NETDEV_XDP_ACT_HW_OFFLOAD));
+	printf("NETDEV_XDP_ACT_RX_SG:\t\t%s\n",
+	       CHECK_XDP_FEATURE(NETDEV_XDP_ACT_RX_SG));
+	printf("NETDEV_XDP_ACT_NDO_XMIT_SG:\t%s\n",
+	       CHECK_XDP_FEATURE(NETDEV_XDP_ACT_NDO_XMIT_SG));
```
Comment (tohojo) [inline xdp-loader/xdp-loader.c:413]: What about future unknown features? At a minimum we should detect that there are unknown bits set and print a notice...

**Review by tohojo — CHANGES_REQUESTED** [review]
A few comments. Also, CI is failing, PTAL...

File: xdp-loader/xdp-loader.8 [hunk 6]
```diff
@@ -160,6 +161,44 @@ Enable debug logging. Specify twice for even more verbosity.
...
+Currently supported XDP features are:
+
+.SS "NETDEV_XDP_ACT_BASIC"
+.PP
+XDP feautues set supported by all drivers (XDP_ABORTED, XDP_DROP, XDP_PASS,
+XDP_TX).
+
+.SS "NETDEV_XDP_ACT_REDIRECT"
+.PP
+The netdev supports XDP_REDIRECT.
```
Comment (tohojo) [inline xdp-loader/xdp-loader.8:178]: "The network device supports handling the XDP_REDIRECT return code. This means packets can be redirected *from* this device by XDP."

File: xdp-loader/xdp-loader.8 [hunk 7]
```diff
@@ -160,6 +161,44 @@ Enable debug logging. Specify twice for even more verbosity.
...
+
+.SS "NETDEV_XDP_ACT_NDO_XMIT"
+.PP
+This feature informs if netdev implements ndo_xdp_xmit callback.
```
Comment (tohojo) [inline xdp-loader/xdp-loader.8:182]: "The networking interfaces implements the ndo_xdp_xmit callback. This means packets can be redirected *to* this device by XDP."

File: xdp-loader/xdp-loader.8 [hunk 8]
```diff
@@ -160,6 +161,44 @@ Enable debug logging. Specify twice for even more verbosity.
...
+
+.SS "NETDEV_XDP_ACT_XSK_ZEROCOPY"
+.PP
+This feature informs if netdev supports AF_XDP in zero copy mode.
```
Comment (tohojo) [inline xdp-loader/xdp-loader.8:186]: "The networking device supports..." (let's be consistent with the phrasing)

File: xdp-loader/xdp-loader.8 [hunk 9]
```diff
@@ -160,6 +161,44 @@ Enable debug logging. Specify twice for even more verbosity.
...
+
+.SS "NETDEV_XDP_ACT_HW_OFFLOAD"
+.PP
+This feature informs if netdev supports XDP hw offloading.
```
Comment (tohojo) [inline xdp-loader/xdp-loader.8:190]: "The networking device supports..."

**Review by tohojo — CHANGES_REQUESTED** [review]
A few more nits, mostly UI things...

**Review by tohojo — CHANGES_REQUESTED** [review]
Code LGTM with a small nit in the test script. However, I totally missed that you were editing the man page file directly. That won't work: the man page is generated by the Makefile from `README.org`, so the changes have to go into that, and then the man page should be regenerated.

File: xdp-loader/tests/test-xdp-loader.sh [hunk 10]
```diff
@@ -90,6 +90,32 @@ test_load_clobber()
     check_run $XDP_LOADER unload $NS --all -vv
 }
 
+check_xdp_feature()
+{
+    check_run ip link add dev v0 type veth peer name v1
+
+    $XDP_LOADER features "$1" | grep "$2" | grep -q "$3"
```
[the remark this answers is not in the document]
Comment (LorenzoBianconi) [reply xdp-loader/tests/test-xdp-loader.sh:97]: @tohojo ack, I will fix it

**Comment by LorenzoBianconi** [comment]
> Code LGTM with a small nit in the test script. However, I totally missed that you were editing the man page file directly. That won't work: the man page is generated by the Makefile from `README.org`, so the changes have to go into that, and then the man page should be regenerated.

@tohojo ack, I will fix it

**Review by tohojo — APPROVED** [review]

**Review by tohojo — CHANGES_REQUESTED** [review]

**Review by tohojo — APPROVED** [review]

**Pull request merged**